drivers/regulator/regulator_pca9422.c

```diff
@@ -0,0 +1,1234 @@
+	 */
+	for (size_t i = 0U; i < PCA9422_NUM_MODES; i++) {
+
+		/* disable if config mode_uv == 0 */
+		if (config->modes_uv[i] == 0) {
+			/*
+			 * Don't allow disabling of LDO1, which is supposed
+			 * to be always on. there is a disable bit, but
+			 * it applies to all modes, and we don't want to
+			 * bother checking that the device tree properly
+			 * configured all modes to be off or all on.
+			 */
+			if (config->desc->enable_reg == PCA9422_LDO1_CFG2) {
+				LOG_ERR("Not disabling always on LDO");
+				return -ENOTSUP;
+			}
+			if (enmode_val == 0) {
+				/* Set enmode_val so the first mode
+				 * that has been disabled.
+				 */
+				enmode_val = 4 - i;
+			}
+		} else if (config->modes_uv[i] > 0) {
+			if (enmode_val) {
+				LOG_ERR("Not possible to have regulator "
+					"enabled in mode %u when configured"
+					" to be disabled in a lower mode", i);
+				return -ENOTSUP;
+			}
+		}
+	}
+
+	/* have to unlock first or else register read returns 0x0 */
+	ret = regulator_pca9422_unlock(&cconfig->i2c);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (enmode_val) {
+		/* Configuration has this regulator disabled in some modes. */
+		if (enmode_val == PCA9422_ENMODE_ENABLED_NONE) {
+			/* Disabling in all modes can only be done via enable_reg */
+			ret = i2c_reg_update_byte_dt(&cconfig->i2c,
+						     config->desc->enable_reg,
+						     config->desc->enable_mask, 0U);
+		} else {
+			/*
+			 * Disabling in non ACTIVE modes can be done via
+			 * the mode_reg.
+			 */
+			ret = i2c_reg_update_byte_dt(&cconfig->i2c,
+						     config->desc->mode_reg,
+						     config->desc->enmode_mask,
+						     enmode_val << config->desc->enmode_pos);
+		}
+
+		if (ret < 0) {
+			goto out;
+		}
+	}
+
+	/* configure mode voltages */
+	for (size_t i = 0U; i < ARRAY_SIZE(config->modes_uv); i++) {
+		if (config->modes_uv[i] > 0) {
+			uint16_t idx;
+
+			/* program mode voltage */
+			ret = linear_range_group_get_win_index(
+				config->desc->ranges, config->desc->num_ranges,
+				config->modes_uv[i], config->modes_uv[i], &idx);
+			if (ret == -EINVAL) {
+				LOG_ERR("invalid voltage %u", config->modes_uv[i]);
+				goto out;
+			}
+
+			ret = i2c_reg_update_byte_dt(
+				&cconfig->i2c,
+				config->desc->vsel_mode[i].vsel_reg,
+				config->desc->vsel_mode[i].vsel_mask, (uint8_t)idx);
+			if (ret < 0) {
+				goto out;
+			}
+		}
+	}
+
+	/* configure lpmode */
+	if (config->desc->lpmode_mask) {
+		ret = i2c_reg_update_byte_dt(&cconfig->i2c,
+					     config->desc->mode_reg,
+					     config->desc->lpmode_mask,
+					     config->lpmode << config->desc->lpmode_pos);
+		if (ret < 0) {
+			goto out;
+		}
+	} else {
+		/* LDO1 and LDO4 have no LPMODE control register. Value should be 3. */
+		assert(config->lpmode == 3);
+	}
+
+out:
+	regulator_pca9422_lock(&cconfig->i2c);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return regulator_common_init(dev, false);
+}
+
+int regulator_pca9422_dvs_state_set(const struct device *dev,
```

Comment:
I think we should consider mapping DVS modes like so:

| Mode Index | Name |
|------------|------|
| 0          | DVS0 |
| 1          | DVS1 |
| 2          | DVS2 |
| 3          | DVS3 |
| 4          | DVS4 |
| 5          | DVS5 |
| 6          | DVS6 |
| 7          | DVS7 |
| 8          | SLEEP |
| 9          | STANDBY |

I also think we should not permit editing the voltage settings for DPSTANDBY mode- effectively DPSTANDBY and SLEEP are the same mode, except outputs can be disabled and LP modes can be used depending on which mode the regulator is in.

Reply:
As I mentioned, I was not trying to introduce/implement new features in the PCA9422 driver that the PCA9420 did not have. Just enough for bringup and sleep mode.

There are other features, like charging, fuel_gauge library (AMUX), that aren't yet implemented too.

Reply:
Sure- I want to be clear, I have no issue with excluding the DVS support from this PR. I just want to make sure that the *external* interface to this driver (DTS, KConfig, C API functions) does not change whenever DVS support is desired)

Reply:
I'd copied the dvs API from the PCA9420 driver, though I think it was rather ackward. It appears to use the dvs APIs not as a way to change dvs modes (which I think was more of the intention), but just as a way to query at runtime the device tree configuration for the voltages of the four modes.

I don't see a lot of value in that, so maybe we should just not implement the dvs APIs for the PCA9422 for now?

Reply:
I agree that omitting the DVS stuff from this initial PR is a good idea, especially if this is subject to change for a future silicon revision.